db/sql/rounds/check_snapshot_availability.sql

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright © 2018 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+
+/*
+  DESCRIPTION: Check round snapshot availability for a particular round
+               Returns TRUE when snapshot available or table is empty, FALSE otherwise
+
+  PARAMETERS: round - Round for we are checking availability
+*/
+
+SELECT (
+	CASE WHEN (
+		SELECT 1 AS available FROM mem_round_snapshot WHERE round = ${round} LIMIT 1
+	) = 1 THEN TRUE
+	ELSE (
+		CASE WHEN (
+			SELECT COUNT(*) FROM mem_round_snapshot
```

Comment:
This case of query is for when we rollback to first block?

Reply:
wouldn't it be nicer to just write to mem_round_snapshot and mem_votes_snapshot after first block?

Reply:
I think this query is overcomplicated. We are adding this case specifically for the first round. So, we can modify the function that calls this query to handle that edge case (for first round), IMO we shouldn't move this logic to the database layer.

Reply:
It's not for the first round, it can happen in every round. There are 2 options:
- one query like this, that returns `true` or `false` if we can use snapshot
- 2 queries
-- one that give us info if snapshot table contains records for requested round
-- second that counts records in snapshot table
and perform condition in round logic:
```
	checkSnapshotAvailability() {
		return this.t.rounds
			.isRoundExistsInRoundSnapshot(this.scope.round)
			.then(isExists => {
				if (!isExists) {
					return this.t.rounds
						.countRoundSnapshot()
						.then(count => {
							if (count) {
								throw new Error(
									`Snapshot for round ${
										this.scope.round
										} not available, unable to perform round rollback`
								);
							}
						});
				}
			});
	}
```
For me first option is better because we have:
- less queries in db repo (don't need additional file + additional function + unit tests for it)
- less complicated logic and unit tests for checkSnapshotAvailability function

Reply:
As you said,
> It's not the best solution, but not so hard to implement and reliable.
this might not be final solution, so this might not be applicable,
- complicated SQL is harder to optimize the performance later
- being able / easy to test in application is better than harder to test on DB
- 1 complex query vs 2 simple query, I think 2 simple query would be better

Reply:
- There are only 2 simple conditions in that query, is it really complicated?
- It's not hard to test that query on DB, tests for that are part of this PR
- Regarding performance - it should be faster than executing 2 queries, and I don't think that there is a room for further performance optimization later.

Reply:
By `not the best solution` I meant there are some better that require a little brainstorming, for example:
- decouple `tick` to `beforeTick` and `afterTick` (so we can snapshot before last block of round, not after one block before last block of round)
- rewrite round 😉

Reply:
I guess I'm just strongly against putting the logic in the database layer.
- 2 simple select can be optimize a lot easier than 1 sql with condition.
- it's always a lot harder to optimize database layer compare to application layer

Even though this 1 SQL would be performance wise slightly better, for consistency through out the code and future scalability, I think we should put it on application layer.

Reply:
@shuse2 Ok, I will rewrite.

Reply:
You guys are thinking this query complicated? You had not seen others queries already. 

Complicated SQL is a relative term, completely depending on the nature of the application you are dealing with. If you are writing a simple e-commerce system, then yes any query more than 2/2 lines or with sub queries can be considered complicated. But if you are dealing with some analytical system, normally a query with 2-3 joins and sub queries considered simple ;) 

Depending on the nature of the application we are dealing, we have to gave some luxury to database layer to handle some complexity. Don't consider DB layer as abstraction to run one line queries.  Testing DB level complex queries is not difficult at all, we already have tests for more complicated db scenarios for this purpose.

Reply:
`SELECT COUNT(*) FROM mem_round_snapshot WHERE round = ?`
can't you just do this and handle the error properly?

Reply:
I would suggest split the query to keep the logic in JS code where possible.

Reply:
No, it will not work, we need to allow restoring snapshot in 2 cases:
- when we have some records with requested round (means we have data for that round, and can use it)
- when we don't have any records for requested round and table is empty (table is empty only when there was no transactions during entire round)